pkg/plugin/v1/init.go

```diff
@@ -0,0 +1,190 @@
+
+	"sigs.k8s.io/kubebuilder/internal/cmdutil"
+	"sigs.k8s.io/kubebuilder/internal/config"
+	"sigs.k8s.io/kubebuilder/pkg/plugin"
+	"sigs.k8s.io/kubebuilder/pkg/plugin/internal"
+	"sigs.k8s.io/kubebuilder/pkg/scaffold"
+)
+
+type initPlugin struct { // nolint:maligned
+	config *config.Config
+
+	// boilerplate options
+	license string
+	owner   string
+
+	// deprecated flags
+	depFlag *pflag.Flag
+	depArgs []string
+	dep     bool
+
+	// flags
+	fetchDeps          bool
+	skipGoVersionCheck bool
+}
+
+var _ plugin.Init = &initPlugin{}
+
+func (p initPlugin) UpdateContext(ctx *plugin.Context) {
+	ctx.Description = `Initialize a new project including vendor/ directory and Go package directories.
+
+Writes the following files:
+- a boilerplate license file
+- a PROJECT file with the domain and repo
+- a Makefile to build the project
+- a go.mod with project dependencies
+- a Kustomization.yaml for customizating manifests
+- a Patch file for customizing image for manager manifests
+- a Patch file for enabling prometheus metrics
+- a cmd/manager/main.go to run
+
+project will prompt the user to run 'dep ensure' after writing the project files.
+`
+	ctx.Examples = fmt.Sprintf(`  # Scaffold a project using the apache2 license with "The Kubernetes authors" as owners
+  %s init --project-version=1 --domain example.org --license apache2 --owner "The Kubernetes authors"
+`, ctx.CommandName)
+}
+
+func (p *initPlugin) BindFlags(fs *pflag.FlagSet) {
+	fs.BoolVar(&p.skipGoVersionCheck, "skip-go-version-check",
+		false, "if specified, skip checking the Go version")
+
+	// dependency args
+	fs.BoolVar(&p.fetchDeps, "fetch-deps", true, "ensure dependencies are downloaded")
+
+	// deprecated dependency args
+	fs.BoolVar(&p.dep, "dep", true, "if specified, determines whether dep will be used.")
+	p.depFlag = fs.Lookup("dep")
+	fs.StringArrayVar(&p.depArgs, "depArgs", nil, "additional arguments for dep")
+
+	if err := fs.MarkDeprecated("dep", "use the fetch-deps flag instead"); err != nil {
+		log.Printf("error to mark dep flag as deprecated: %v", err)
+	}
+	if err := fs.MarkDeprecated("depArgs", "will be removed with version 1 scaffolding"); err != nil {
+		log.Printf("error to mark dep flag as deprecated: %v", err)
+	}
+
+	// boilerplate args
+	fs.StringVar(&p.license, "license", "apache2",
+		"license to use to boilerplate, may be one of 'apache2', 'none'")
+	fs.StringVar(&p.owner, "owner", "", "owner to add to the copyright")
+
+	// project args
+	p.config = config.New(config.DefaultPath)
+	fs.StringVar(&p.config.Repo, "repo", "", "name to use for go module (e.g., github.com/user/repo), "+
+		"defaults to the go package of the current working directory.")
+	fs.StringVar(&p.config.Domain, "domain", "my.domain", "domain for groups")
+}
+
+func (p *initPlugin) Run() error {
+	return cmdutil.Run(p)
+}
+
+func (p *initPlugin) SetVersion(v string) {
+	if p.config == nil {
+		p.config = config.New(config.DefaultPath)
+	}
+	p.config.Version = v
+}
```

Comment:
Can `SetVersion` be called before`BindFlags`?
- If the answer is yes, `BindFlags` will make that call be useless as it will create a new config and won't keep the version, you should add a `p.config == nil` there too.
- If the answer is no, this check is useless as `BindFlags` already initializes this field.

Aplies to v2

Reply:
Order doesn't matter as long as `p.config` isn't nil, so I'll remove the check in `SetVersion`.

Reply:
It does matter as BindFlags is creating a new instance of Config, so the stored version is lost in case it is run before, and if it is run after BindFlags no need to check for nil as you already know it won't be nil.